File: glass.py
from hal import init,give_hard,give_soft,go_home 
import time

class Ingredient:
	def __init__(self,name, position, category):
		self.name = name
		self.position = position
		self.category = category

class Dose:
	def __init__(self,ingredient,quantity):
		self.ingredient = ingredient
		self.quantity = quantity

### SOFT INGREDIENT
Coca = Ingredient('Coca',1,'Soft')
Eau = Ingredient('Eau',7,'Soft')
Orange = Ingredient('Orange',2,'Soft')
Ananas = Ingredient('Ananas',3,'Soft')


### HARD INGREDIENT
Rhum = Ingredient('Rhum',2,'Hard')
Tequila = Ingredient('Tequila',3,'Hard')
Grenadine = Ingredient('Grenadine',4,'Hard')
Vodka = Ingredient('Vodka',5,'Hard')
Whiskey = Ingredient('Whiskey',6,'Hard')

### DOSES
D_Whiskey = Dose(Whiskey,5)
D_Rhum = Dose(Rhum,5)
D_Grenadine = Dose(Grenadine,0.01)
D_Tequila = Dose(Tequila,5)
D_Vodka = Dose(Vodka,5)

D_Coca = Dose(Coca,4)
D_Orange = Dose(Orange,4)
D_Ananas = Dose(Ananas,4)
D_Orange_short = Dose(Orange,3)
D_Ananas_short = Dose(Ananas,3)

### RECEIPES
R_Whiskey_Coca = [D_Whiskey,D_Coca]
R_Cuba_Libre = [D_Rhum,D_Coca]
R_Punch = [D_Rhum,D_Orange]
R_Tequila_Sunrise = [D_Grenadine,D_Tequila,D_Orange]
R_Sex_On_The_Beach = [D_Grenadine,D_Vodka,D_Orange]
R_Punch_Planteur = [D_Grenadine,D_Rhum,D_Orange]
R_After_Glow = [D_Grenadine,D_Orange_short,D_Ananas_short]

def emergency_stop():
    print "In glass.py : Emergency STOP"
    init()

def choose_cocktail(cocktail_name):
    print "In choose_cocktail : Will prepare " + cocktail_name
    if(cocktail_name=="whiskycoca"):
        make_cocktail(R_Whiskey_Coca)
    if(cocktail_name=="cubalibre"):
        make_cocktail(R_Cuba_Libre)
    if(cocktail_name=="punch"):
        make_cocktail(R_Punch)
    if(cocktail_name=="tequilasunrise"):
        make_cocktail(R_Tequila_Sunrise)
    if(cocktail_name=="sexonthebeach"):
        make_cocktail(R_Sex_On_The_Beach)
    if(cocktail_name=="afterglow"):
        make_cocktail(R_After_Glow)
    if(cocktail_name=="punchplanteur"):
        make_cocktail(R_Punch_Planteur)

def make_cocktail(receipe):
    print "In make_cocktail : Will do the receipe"
    init()
    for dose in receipe:
        print " Dose : "
        if(dose.ingredient.category == 'Hard'):
            give_hard(dose.ingredient.position, dose.quantity)
        if(dose.ingredient.category == 'Soft'):
            give_soft(dose.ingredient.position,dose.quantity)
        #init()
    #go_home()
    init()

File: glass.lua

Ingredient = {}

function create_ingredient(name, position, category)
    return {
       name=name,
       position=position,
       category=category
    }
end

-- SOFT INGREDIENT
Coca = create_ingredient('Coca',3,'Soft')
Eau = create_ingredient('Eau',7,'Soft')
Orange = create_ingredient('Orange',1,'Soft')
Ananas = create_ingredient('Ananas',2,'Soft')

-- HARD INGREDIENT
Vodka = create_ingredient('Vodka',1,'Hard')
Rhum = create_ingredient('Rhum',2,'Hard')
Tequila = create_ingredient('Tequila',3,'Hard')
Grenadine = create_ingredient('Grenadine',4,'Hard')
Whiskey = create_ingredient('Whiskey',5,'Hard')

Dose = {}

function create_dose(ingredient,quantity)
    return {
	    ingredient = ingredient,
	    quantity = quantity
    }
end

function copy_dose(another)
    return create_dose(another.ingredient, another.quantity)
end

-- doses
D_Rhum = create_dose(Rhum,5)
D_Grenadine = create_dose(Grenadine,0.001)
D_Tequila = create_dose(Tequila,5)
D_Vodka = create_dose(Vodka,5)

D_Coca = create_dose(Coca,4)
D_Orange = create_dose(Orange,4)
D_Ananas = create_dose(Ananas,4)
D_Orange_short = create_dose(Orange,2)
D_Ananas_short = create_dose(Ananas,2)

-- recipes map
recipes = {}
recipes["Whiskey_Coca"] = {D_Whiskey, D_Coca}
recipes["Cuba_Libre"] = {D_Rhum, D_Coca}
recipes["Punch"] = {D_Rhum, D_Orange}
recipes["Tequila_Sunrise"] = {D_Grenadine, D_Tequila, D_Orange}
recipes["Sex_On_The_Beach"] = {D_Grenadine, D_Vodka, D_Orange}
recipes["Punch_Planteur"] = {D_Grenadine, D_Rhum, D_Orange}
recipes["After_Glow"] = {D_Grenadine, D_Orange_short, D_Ananas_short}
recipes["Orange"] = {D_Orange}

function copy_recipe(recipe)
    copy = {}
    for i, dose in ipairs(recipe) do
        table.insert(copy, dose)
    end
    return copy
end

function emergency_stop()
    print("In glass.py : Emergency STOP")
    init()
end

function table_length(table)
    local count = 0
    for _ in pairs(table) do count = count + 1 end
    return count
end

function make_cocktail(original_recipe, give_hard_dose, give_soft_dose, reset, callback)
    -- takes two function wich serve doses based on position
    -- and quantity one for hard, the other for soft
    -- plus a reset for the machine and a callback when done
    local recipe = copy_recipe(original_recipe)
    local function make_cocktail_action(dose, next_actions)
        local function cocktail_action()
	    print("Cocktail Action : "..dose.ingredient.name .. dose.ingredient.category)
            if(dose.ingredient.category == 'Hard') then
                give_hard_dose(dose.ingredient.position, dose.quantity, next_actions)
            end
            if(dose.ingredient.category == 'Soft') then
                give_soft_dose(dose.ingredient.position,dose.quantity, next_actions)
            end
        end
        return cocktail_action
    end
    local function pile_up_recipe(next_actions)
        print('piling...')
        if (table_length(recipe) == 0) then
            return next_actions
        else
            dose = table.remove(recipe)
            print("pile up" .. dose.ingredient.name)
            return pile_up_recipe(make_cocktail_action(dose, next_actions))
        end
    end
    local function cocktail_last_action()
        reset(callback)
    end
    local cocktail_execution = pile_up_recipe(cocktail_last_action)
    cocktail_execution()
end
